man/man1/txt2grid.1
.TH txt2grid 1  "USER COMMANDS"
.SH NAME
txt2grid \- convert text files to clm grid data files for LPJ
.SH SYNOPSIS
.B txt2grid
[\-h] [\-v] [\-latlon] [\-map refgridfile] [-fmt s] [-skip n] [-cellsize size] [-cellsize_lon size] [-cellsize_lat size] [-float] [-double] [-json]
.I txtfile gridfile
.SH DESCRIPTION
Program converts text files into LPJ grid data files
By default text files have to be in the following format:

.nf
Id,Lon,Lat,X_CRU,Y_CRU
1,-179.75,-16.25,1,148
2,..
.ni
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-v,\--verbose
Verbose output.
.TP
\-latlon
read first latitude then longitude value from text file.
.TP
\-map refgridfile
Map cell coordinate to nearest cell in refgridfile.
.TP
\-fmt s
Input format string for the scanf function in
.B txt2grid
, default is "%*d,%f,%f,%*d,%*d".
.TP
\-skip n
Skip first n lines in text file, default is 1.
.TP
\-cellsize size
Set latitudinal and longitudinal cell size, default is 0.5.
.TP
\-cellsize_lon size
Set longitudinal cell size, default is 0.5.
.TP
\-cellsize_lat size
Set latitudinal cell size, default is 0.5.
.TP
\-float
Longitude/latitude values written in float format to clm file, default is short.
.TP
\-double
Longitude/latitude values written in double format to clm file, default is short.
.TP
\-json
JSON metafile is created with suffix '.json'
.I textfile
Filename of text file to convert.
.TP
.I gridfile
Filename of clm grid data file written.
.SH EXAMPLES
.TP
Create clm grid data file from text file containing longitude/latitude values in each line:
.B txt2grid
-skip 0 -fmt "%f %f" grid.txt grid.clm
.TP
Create clm grid data file with double values from text file containing longitude/latitude values in each line:
.B txt2grid
-skip 0 -double -fmt "%lf %lf" grid.txt grid.clm
.PP
.SH EXIT STATUS
.B txt2grid
returns a zero exit status if the clm file is created.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
lpjml(1), txt2clm(1), cru2clm(1), grid2clm(1), catclm(1), cdf2grid(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)

### Added

- Option `-h` added to utilities `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.


## [5.9.25] - 2025-05-22

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Fabian Stenzel (stenzel@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de)

### Changed

- Axis names and missing values for NetCDF output are read from JSON file `netcdf.cjson` instead of being hard coded.
- Warnings in `addheader` replaced by terminating errors.
- Function `openinputfile()` returns an error if cell size differs in JSON metafile.
- Function `openmetafile()` returns an error if binary file is empty.
- Output format for cell size updated in `fprintheader.c`, `fprintoutputjson.c`, `fprintjson.c`, and `openinputfile.c`.

### Added

- New utility `splitclm` added to copy specific bands from a clm file.
- Environment variables `LPJNOPP` and `LPJPEDANTIC` added. If set to `true`, preprocessor is disabled and pedantic checking is enabled, respectively.
- Info about `--help` option added in case of error in command line arguments.
- Option `-config` added to `cdf2clm`, `bin2cdf`, `country2cdf` and `clm2cdf` in order to read axis names and missing values from a JSON file. File `netcdf.json` added as a template.
- Option `-json` added to `arr2clm`, `asc2clm`, `txt2grid`, `txt2clm`, `getcountry`, and `cdf2coord` in order to create additional JSON metafiles.
- Option `-inpath` added to `regridlpj` in order to set directory where data can be found.
- Options `-zero` and `-search` added to `regridlpj`.
- Option `-double` added to `txt2grid` in order to create grid files with coordinate values of type double.
- Option `-timestep` added to `addheader`, `txt2clm` and `bin2cdf`.
- Option `sum`, `tsum` and `tmean` to `mathclm` added.
- Check for correct setting of nstep and nbands added in `txt2clm`.

### Fixed

- Misspelled `-fsanitize=address` option corrected in `Makefile.mpich`.
- Missing `free(cmd)` added in `fprintincludes.c`.
- Parsing of options corrected in `fprintincludes.c`.
- Utility `bin2cdf` fixed for timestep>1.


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ``.gitlab/merge_request_templates/default.md`` that includes several checklists.


## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.

src/utils/txt2grid.c
/**************************************************************************************/
/**                                                                                \n**/
/**                      t  x  t  2  g  r  i  d  .  c                              \n**/
/**                                                                                \n**/
/**     Program converts text data to grid clm file format. Data must be           \n**/
/**     in the following format:                                                   \n**/
/**                                                                                \n**/
/**     Id,Lon,Lat,X_CRU,Y_CRU                                                     \n**/
/**     1,-179.75,-16.25,1,148                                                     \n**/
/**     2,..                                                                       \n**/
/**     or in a format specified by the -fmt option                                \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "\nUsage: txt2grid [-h] [-v] [-map file] [-fmt s] [-skip n] [-cellsize size] [-cellsize_lon size] [-cellsize_lat size]\n       [-float] [-double] [-latlon] [-json] gridfile clmfile\n"
#define ERR_USAGE USAGE "\nTry \"txt2grid --help\" for more information.\n"

typedef  struct
{
  float lon,lat;
} Float_coord;

typedef  struct
{
  double lon,lat;
} Double_coord;

static Bool scancoord(FILE *file,const char *fmt,Double_coord *coord,Bool latlon,Type type)
{
  Float_coord fcoord;
  char *pos,clon,clat;
  if((pos=strstr(fmt,"%c"))!=NULL && strstr(pos,"%c")!=NULL)
  {
    if(latlon)
    {
      if(type!=LPJ_DOUBLE)
      {
        if(fscanf(file,fmt,&fcoord.lat,&clat,&fcoord.lon,&clon)!=4)
          return FALSE;
        coord->lat=fcoord.lat;
        coord->lon=fcoord.lon;
      }
      else if(fscanf(file,fmt,&coord->lat,&clat,&coord->lon,&clon)!=4)
        return FALSE;
    }
    else
    {
      if(type!=LPJ_DOUBLE)
      {
        if(fscanf(file,fmt,&fcoord.lon,&clon,&fcoord.lat,&clat)!=4)
          return FALSE;
        coord->lat=fcoord.lat;
        coord->lon=fcoord.lon;
      }
      else if(fscanf(file,fmt,&coord->lon,&clon,&coord->lat,&clat)!=4)
        return FALSE;
    }
    if(clon=='W')
      coord->lon= -coord->lon;
    else if(clon!='E')
    {
      fprintf(stderr,"Invalid character '%c' for longitude.\n",clon);
      return FALSE;
    }
    if(clat=='S')
      coord->lat= -coord->lat;
    else if(clat!='N')
    {
      fprintf(stderr,"Invalid character '%c' for latitude.\n",clat);
      return FALSE;
    }
    return TRUE;
  }
  else
  {
    if(type!=LPJ_DOUBLE)
    {
      if(fscanf(file,fmt,(latlon) ? &fcoord.lat : &fcoord.lon,(latlon) ? &fcoord.lon : &fcoord.lat)!=2)
        return FALSE;
      coord->lat=fcoord.lat;
      coord->lon=fcoord.lon;
      return TRUE;
    }
    else
     return (fscanf(file,fmt,(latlon) ? &coord->lat : &coord->lon,(latlon) ? &coord->lon : &coord->lat)==2);
  }
} /* of 'scancoord' */

int main(int argc,char **argv)
{
  FILE *file,*gridfile;
  Coordfile coordfile=NULL;
  Filename filename;
  const char *fmt;
  Coord grid,*grid_ref=NULL,res;
  String line;
  int i,iarg,nskip,n,index;
  Double_coord dcoord;
  Float_coord coord;
  Real dist_min;
  Header header;
  char *endptr,*map_name;
  char *arglist,*out_json;
  Bool latlon,verbose,isjson;
  fmt="%*d,%f,%f,%*d,%*d";
  Type type;
  nskip=1;
  header.cellsize_lon=header.cellsize_lat=0.5;
  type=LPJ_SHORT;
  latlon=verbose=isjson=FALSE;
  map_name=NULL;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("   txt2grid (" __DATE__ ") Help\n"
               "   ===========================\n\n"
               "Convert text file to clm grid file for LPJmL version %s\n\n"
               USAGE
               "\nArguments:\n"
               "-h,--help          print this help text\n"
               "-v,--verbose       verbose output\n"
               "-map file          mapping to nearest cell in grid file\n"
               "-fmt s             format string for text input, default is '%s'\n"
               "-cellsize size     cell size, default is %g\n"
               "-cellsize_lat size latitudinal cell size, default is %g\n"
               "-cellsize_lon size longitudinal cell size, default is %g\n"
               "-float             write float data, default is short\n"
               "-double            write double data, default is short\n"
               "-skip n            skip first n lines, default is one\n"
               "-latlon            read latitude then longitude\n"
               "-json              JSON metafile is created with suffix '.json'\n"
               "gridfile           filename of grid text file\n"
               "clmfile            filename of clm data file\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               getversion(),fmt,header.cellsize_lon,header.cellsize_lat,header.cellsize_lon);
        return EXIT_SUCCESS;
      }
      if(!strcmp(argv[iarg],"-fmt"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-fmt' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        fmt=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-v") || !strcmp(argv[iarg],"--verbose"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-float"))
        type=LPJ_FLOAT;
      else if(!strcmp(argv[iarg],"-double"))
        type=LPJ_DOUBLE;
      else if(!strcmp(argv[iarg],"-latlon"))
        latlon=TRUE;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-cellsize' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        header.cellsize_lon=header.cellsize_lat=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for '-cellsize' option.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-cellsize_lat"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-cellsize_lat' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        header.cellsize_lat=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for '-cellsize_lat' option.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-cellsize_lon"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-cellsize_lon' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        header.cellsize_lon=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for '-cellsize_lon' option.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-map"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-map' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        map_name=argv[++iarg];
      }
      else if(!strcmp(argv[iarg],"-skip"))
      {
        if(iarg==argc-1)
        {
           fputs("Argument missing after '-skip' option.\n"
                 ERR_USAGE,stderr);
           return EXIT_FAILURE;
        }
        nskip=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid number '%s' for '-skip' option.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Error: invalid option '%s'.\n"
                ERR_USAGE,argv[iarg]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fputs("Filenames missing.\n"
          ERR_USAGE,stderr);
    return EXIT_FAILURE;
  }
  if(map_name!=NULL)
  {
     filename.name=map_name;
     filename.fmt=CLM;
     coordfile=opencoord(&filename,TRUE);
     if(coordfile==NULL)
       return EXIT_FAILURE;
     grid_ref=newvec(Coord,numcoord(coordfile));
     getcellsizecoord(&header.cellsize_lon,&header.cellsize_lat,coordfile);
     res.lon=header.cellsize_lon;
     res.lat=header.cellsize_lat;
     for(i=0;i<numcoord(coordfile);i++)
       if(readcoord(coordfile,grid_ref+i,&res))
       {
         fprintf(stderr,"Error reading coordinate %d in '%s'.\n",i+1,map_name);
         return EXIT_FAILURE;
       }
  }
  file=fopen(argv[iarg],"r");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s.\n",argv[iarg],strerror(errno));
    return EXIT_FAILURE;
  }
  for(n=0;n<nskip;n++)
    if(fgets(line,STRING_LEN,file)==NULL)
    {
      fprintf(stderr,"Error skipping file header in '%s'.\n",argv[iarg]);
      return EXIT_FAILURE;
    }
  header.ncell=0;
  header.nbands=2;
  header.order=CELLYEAR;
  header.firstcell=0;
  header.firstyear=0;
  header.nyear=1;
  header.nstep=1;
  header.timestep=1;
  header.datatype=type;
  if(type==LPJ_SHORT)
  {
    header.scalar=0.01;
  }
  else
  {
    header.scalar=1;
  }
  gridfile=fopen(argv[iarg+1],"wb");
  if(gridfile==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[iarg+1],strerror(errno));
    return EXIT_FAILURE;
  }
  fwriteheader(gridfile,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  while(scancoord(file,fmt, &dcoord,latlon,type))
  {
    if(map_name!=NULL)
    {
      grid.lon=dcoord.lon;
      grid.lat=dcoord.lat;
      if(verbose)
      {
        fputs("Coordinate ",stdout);
        printcoord(&grid);
        fputs(" mapped to ",stdout);
      }
      index=findnextcoord(&dist_min,&grid,grid_ref,numcoord(coordfile));
      if(verbose)
      {
        printcoord(grid_ref+index);
        printf(", distance=%g\n",dist_min);
      }
      dcoord.lon=grid_ref[index].lon;
      dcoord.lat=grid_ref[index].lat;
    }
    else if(verbose)
    {
      grid.lon=dcoord.lon;
      grid.lat=dcoord.lat;
      printcoord(&grid);
      putchar('\n');
    }
    switch(type)
    {
       case LPJ_FLOAT:
         coord.lat=(float)dcoord.lat;
         coord.lon=(float)dcoord.lon;
         fwrite(&coord,sizeof(coord),1,gridfile);
         break;
       case LPJ_DOUBLE:
         fwrite(&dcoord,sizeof(dcoord),1,gridfile);
         break;
       default:
         grid.lon=dcoord.lon;
         grid.lat=dcoord.lat;
         if(fabs(coord.lon*100-round(coord.lon*100))>1e-3)
           fprintf(stderr,"Warning: Longitude of %.6g at %d cannot be represented by short value of %g.\n",
                   coord.lon,header.ncell,round(coord.lon*100));
         if(fabs(coord.lat*100-round(coord.lat*100))>1e-3)
           fprintf(stderr,"Warning: Latitude of %.6g at %d cannot be represented by short value of %g.\n",
                   coord.lat,header.ncell,round(coord.lat*100));
         writecoord(gridfile,&grid);
    } /* of switch(type) */
    header.ncell++;
  }
  if(header.ncell==0)
    fprintf(stderr,"Error: No grid cells written to '%s'.\n",argv[iarg+1]);
  else
    printf("Number of grid cells: %d\n",header.ncell);
  rewind(gridfile);
  fwriteheader(gridfile,&header,LPJGRID_HEADER,LPJGRID_VERSION);
  fclose(gridfile);
  if(isjson)
  {
    /* write JSON metafile */
    out_json=malloc(strlen(argv[iarg+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[iarg+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    gridfile=fopen(out_json,"w");
    if(gridfile==NULL)
    {
      printfcreateerr(out_json);
      return EXIT_FAILURE;
    }
    fprintjson(gridfile,argv[iarg+1],NULL,"txt2grid",NULL,arglist,&header,NULL,NULL,NULL,0,"grid","degree",NULL,"cell coordinates",NULL,LPJ_SHORT,CLM,LPJGRID_HEADER,FALSE,LPJGRID_VERSION);
    fclose(gridfile);
  }
  if(map_name!=NULL)
    closecoord(coordfile);
  fclose(file);
  return EXIT_SUCCESS;
} /* of 'main' */
